src/python.cpp

```diff
@@ -2415,3 +2416,45 @@ PyObjectRef r_convert_dataframe(RObject dataframe, bool convert) {
   return py_ref(dict.detach(), convert);
 
 }
+
+// [[Rcpp::export]]
+PyObjectRef r_convert_date(DateVector date_vector, bool convert) {
+  
+  PyObjectPtr datetime(PyImport_ImportModule("datetime"));
+  
+  // scalar
+  if (date_vector.size() == 1) {
+    
+    Date date = date_vector[0];
+    PyObjectPtr py_date(PyObject_CallMethod(
+        datetime, "date", "iii",
+        static_cast<int>(date.getYear()),
+        static_cast<int>(date.getMonth()),
+        static_cast<int>(date.getDay())));
+    if (py_date == NULL) {
+      stop(py_fetch_error());
+    }
+    return py_ref(py_date, convert);
+    
+  // vector  
+  } else {
+    
+    PyObjectPtr list(PyList_New(date_vector.size()));
+    for (int i = 0; i < date_vector.size(); ++i) {
+      Date date = date_vector[i];
+      PyObjectPtr py_date(PyObject_CallMethod(
+          datetime, "date", "iii",
+          static_cast<int>(date.getYear()),
+          static_cast<int>(date.getMonth()),
+          static_cast<int>(date.getDay())));
+      if (py_date == NULL) {
+        stop(py_fetch_error());
+      }
+      int res = PyList_SetItem(list, i, py_date);
```

Comment:
`PyList_SetItem` "steals" a reference, and so the associated `py_date` item should either not be wrapped in `PyObjectPtr`, or explicitly detached. See for example:

https://github.com/rstudio/reticulate/blob/64f685ab279516250ef15f2957820744c9841d32/src/python.cpp#L1309-L1313

Reply:
(This is one of the frustrating bits of the Python API: most functions return new references, but some 'borrow' or 'steal' a reference and so have implications on how reference counting is performed)

Reply:
Oh sorry, I should have connected the dots ... (had read about e.g. `PyList_SetItem()` and also seen `PyObjectPtr*` in the piece of code I was taking as an example...)